### PR TITLE
Onboarding Improvements

### PR DESCRIPTION
**Pull request by ReadeP**
## Overview

Improved the onboarding flow for new users and removed it from existing users.

## Changes Made

- Existing users no longer have to onboard again each time they sign out.
- Users can now manually input their location, instead of relying on the poor location api we were using before.
- Users can now select "Grad Student" as an option while onboarding.
- Users can now select between 1 and 3 prompts to complete while onboarding.
- Fixed a few instances where the keyboard couldn't be dismissed and users would be stuck on an onboarding screen.
- Fixed a few miscellaneous warnings.

## Test Coverage

Tested manually on device.

## Next Steps

- Prompts are not showing on profile and there are a few messaging keyboard/scrollview bugs that need to be addressed.
- Image uploading does not currently work.
- Backend needs to add support for users filling out less than 3 prompts and select "Grad Student". Right now we get an 500 error in the app.

### CODE REVIEW

File: Pear/Controller/LoginViewController.swift [hunk 1]
```diff
@@ -111,16 +111,31 @@ class LoginViewController: UIViewController {
...
+            guard let user = try? result.get(), let hasOnboarded = user.hasOnboarded, hasOnboarded else {
+                let onboardingVC = OnboardingPageViewController(transitionStyle: .scroll, navigationOrientation: .horizontal)
+                self.navigationController?.pushViewController(onboardingVC, animated: true)
+                return
+            }
+            
+            self.navigationController?.pushViewController(HomeViewController(), animated: true)
+        }
+    }
+
```
Comment (Npikielny) [inline Pear/Controller/LoginViewController.swift:139]: Two empty lines next to each other

File: Pear/View/OnboardingSelectDropdownView.swift [hunk 2]
```diff
@@ -129,7 +128,11 @@ extension OnboardingSelectDropdownView: UITableViewDelegate {
     func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
         // Updates dropdown text when a cell is selected in the table view and hides the table view.
         let selectedText = tableData[indexPath.row]
-        dropdownButton.setTitle("\(textTemplate) \(selectedText)", for: .normal)
+        if selectedText == "Grad Student" {
+            dropdownButton.setTitle("\(selectedText)", for: .normal)
+        } else {
+            dropdownButton.setTitle("\(textTemplate) \(selectedText)", for: .normal)
+        }
```
Comment (Npikielny) [inline Pear/View/OnboardingSelectDropdownView.swift:135]: could make this ternary ¯\_(ツ)_/¯

Comment (Npikielny) [reply Pear/View/OnboardingSelectDropdownView.swift:135]: Same with 112

File: Pear/View/OnboardingSearchDropdownView.swift [hunk 3]
```diff
@@ -201,19 +181,36 @@ extension OnboardingSearchDropdownView: UISearchBarDelegate, UITableViewDelegate
...
     }
+    
+    func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
+        searchBar.searchTextField.resignFirstResponder()
+        tableView.isHidden = true
+        delegate?.sendDropdownViewToBack(dropdownView: self)
+        
+        guard searchType == .places else {
+            return
+        }
```
Comment (Npikielny) [inline Pear/View/OnboardingSearchDropdownView.swift:208]: Could change to one line

File: Pear/Controller/LoginViewController.swift [hunk 4]
```diff
@@ -111,16 +111,31 @@ class LoginViewController: UIViewController {
...
             }
         }
     }
+    
+    private func getUser() {
+        NetworkManager.getMe { [weak self] result in
+            guard let self = self else { return }
+            
+            // If the user already exists and has onboarding, we push them directly to the HomeViewController.
+
```
Comment (Npikielny) [inline Pear/Controller/LoginViewController.swift:128]: delete extra line

**Review by Npikielny — APPROVED** [review]
Looks good to me 🤩🤩🚀🚀

File: Pear/Controller/LoginViewController.swift [hunk 5]
```diff
@@ -111,16 +111,31 @@ class LoginViewController: UIViewController {
...
                 case .failure(let error):
                     print(error.localizedDescription)
                 }
             }
         }
     }
+    
+    private func getUser() {
+        NetworkManager.getMe { [weak self] result in
+            guard let self = self else { return }
```
Comment (Npikielny) [inline Pear/Controller/LoginViewController.swift:125]: What does this do?

File: Pear/View/OnboardingSelectDropdownView.swift (hunk 2 above)
Comment (ReadeP) [reply Pear/View/OnboardingSelectDropdownView.swift:135]: I think so too

`
let title = selectedText == "Grad Student" ? selectedText : "\(textTemplate) \(selectedText)"
dropdownButton.setTitle(title)
`

File: Pear/Controller/LoginViewController.swift [hunk 6]
```diff
@@ -111,16 +111,31 @@ class LoginViewController: UIViewController {
...
                     }
                 case .failure(let error):
                     print(error.localizedDescription)
                 }
             }
         }
     }
+    
+    private func getUser() {
+        NetworkManager.getMe { [weak self] result in
```
Comment (danielVebman) [inline Pear/Controller/LoginViewController.swift:124]: `getUser` or `getCurrentUsed`seems like a better name. `getMe` feels informal. Maybe not.

File: Pear/Controller/LoginViewController.swift [hunk 7]
```diff
@@ -111,16 +111,31 @@ class LoginViewController: UIViewController {
...
+            guard let self = self else { return }
+            
+            // If the user already exists and has onboarding, we push them directly to the HomeViewController.
```
Comment (danielVebman) [inline Pear/Controller/LoginViewController.swift:127]:
```suggestion
            // If the user already exists and has onboarded, we push them directly to the HomeViewController.
```

File: Pear/Controller/LoginViewController.swift [hunk 8]
```diff
@@ -111,16 +111,31 @@ class LoginViewController: UIViewController {
...
+            
+            guard let user = try? result.get(), let hasOnboarded = user.hasOnboarded, hasOnboarded else {
+                let onboardingVC = OnboardingPageViewController(transitionStyle: .scroll, navigationOrientation: .horizontal)
+                self.navigationController?.pushViewController(onboardingVC, animated: true)
+                return
+            }
+            
+            self.navigationController?.pushViewController(HomeViewController(), animated: true)
```
Comment (danielVebman) [inline Pear/Controller/LoginViewController.swift:135]: Does this run on the main thread?

File: Pear/Onboarding/AnswerPromptViewController.swift [hunk 9]
```diff
@@ -179,6 +180,7 @@ extension AnswerPromptViewController: UITextViewDelegate {
     }
 
     func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
+        if text == "\n" { textView.resignFirstResponder() }
```
Comment (danielVebman) [inline Pear/Onboarding/AnswerPromptViewController.swift:183]: There's [`textFieldShouldReturn`](https://developer.apple.com/documentation/uikit/uitextfielddelegate/1619603-textfieldshouldreturn)

File: Pear/Onboarding/ProfilePromptsViewController.swift [hunk 10]
```diff
@@ -107,7 +107,7 @@ class ProfilePromptsViewController: UIViewController {
     }
 
     private func updateNext() {
-        nextButton.isEnabled = prompts.filter{ $0.answer != nil }.count == 3
+        nextButton.isEnabled = prompts.filter{ $0.answer != nil }.count > 0
```
Comment (danielVebman) [inline Pear/Onboarding/ProfilePromptsViewController.swift:110]:
```suggestion
        nextButton.isEnabled = !prompts.compactMap(\.answer).isEmpty
```

File: Pear/View/OnboardingSelectDropdownView.swift [hunk 11]
```diff
@@ -129,7 +128,11 @@ extension OnboardingSelectDropdownView: UITableViewDelegate {
     func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
         // Updates dropdown text when a cell is selected in the table view and hides the table view.
         let selectedText = tableData[indexPath.row]
-        dropdownButton.setTitle("\(textTemplate) \(selectedText)", for: .normal)
+        if selectedText == "Grad Student" {
+            dropdownButton.setTitle("\(selectedText)", for: .normal)
```
Comment (danielVebman) [inline Pear/View/OnboardingSelectDropdownView.swift:132]:
```suggestion
            dropdownButton.setTitle(selectedText, for: .normal)
```

File: Pear/View/OnboardingSelectDropdownView.swift (hunk 2 above)
Comment (danielVebman) [reply Pear/View/OnboardingSelectDropdownView.swift:135]: Idk if I would make this ternary. This is pretty readable. Ternary expressions are best with short values.

**Review by danielVebman — APPROVED** [review]
Lgtm 🍐🚀

File: Pear/Onboarding/AnswerPromptViewController.swift (hunk 9 above)
Comment (ReadeP) [reply Pear/Onboarding/AnswerPromptViewController.swift:183]: This is a textview which doesn't have a similar delegate func

File: Pear/Onboarding/DemographicsViewController.swift [hunk 12]
```diff
@@ -65,12 +65,16 @@ class DemographicsViewController: UIViewController {
         subtitleLabel.font = ._12CircularStdBook
         view.addSubview(subtitleLabel)
 
+        // TODO: Add Grad student option
+
```
Comment (haiyingweng) [inline Pear/Onboarding/DemographicsViewController.swift:69]: should this be removed?

File: Pear/View/OnboardingSelectDropdownView.swift [hunk 13]
```diff
@@ -109,15 +109,14 @@ class OnboardingSelectDropdownView: UIView {
 
     /// Set text of field if value already exists
     func setSelectValue(value: String) {
-        dropdownButton.setTitle("\(textTemplate) \(value)", for: .normal)
+        if value == "Grad Student" {
```
Comment (haiyingweng) [inline Pear/View/OnboardingSelectDropdownView.swift:112]: since "Grad Student" appears in more than 1 place, might be better to make it a variable?

**Review by haiyingweng — APPROVED** [review]
lgtm!

File: Pear/Controller/LoginViewController.swift (hunk 6 above)
Comment (ReadeP) [reply Pear/Controller/LoginViewController.swift:124]: yeah getMe is wacky

**Pull request merged**